pkg/scheduler/plugins/reservation/hook.go

```diff
@@ -0,0 +1,209 @@
+/*
+Copyright 2022 The Koordinator Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package reservation
+
+import (
+	"context"
+	"fmt"
+
+	corev1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/labels"
+	quotav1 "k8s.io/apiserver/pkg/quota/v1"
+	"k8s.io/klog/v2"
+	resourceapi "k8s.io/kubernetes/pkg/api/v1/resource"
+	"k8s.io/kubernetes/pkg/scheduler/framework"
+
+	"github.com/koordinator-sh/koordinator/pkg/scheduler/frameworkext"
+	"github.com/koordinator-sh/koordinator/pkg/util"
+)
+
+var (
+	_ frameworkext.PreFilterPhaseHook = &Hook{}
+	_ frameworkext.FilterPhaseHook    = &Hook{}
+)
+
+type Hook struct{}
+
+func NewHook() *Hook {
+	return &Hook{}
+}
+
+func (h *Hook) Name() string { return Name }
+
+func (h *Hook) PreFilterHook(handle frameworkext.ExtendedHandle, cycleState *framework.CycleState, pod *corev1.Pod) (*corev1.Pod, bool) {
```

Comment:
As discussed with @allwmh before, do we need PreFilter Hook? Or we can consider implementing this ability by adjusting the order of the PreFilter plugins.

Reply:
I think PreFilter Hook is needed.  
To support reservation, we alter the pod copy and pass it to PreFilter plugins, while these plugins use the Pod object from PodInformer originally. If we turned to re-order the execution of PreFilter plugins and change the pod object inside Pre-Filter, the initial object would get polluted.

Reply:
OK. got it!